### ISSUE TITLE
Support mutable render object and cache render result

### ISSUE THREAD

**Issue by qiumuyang** [iori2333/render#7]
Caching can be useful when we want to repeat some objects in a `Container`, `Stack`, etc.

---

But caching comes with problems. If some attribute of an object is modified, the rendered image should update accordingly instead of reading from the cache. 

- A straightforward way is to use `lru_cache` and call `cache_clear` explicitly each time an object is modified, which is inefficient and not elegant. 
- Another solution is to forbid "re-render" of an object. Object attributes cannot be set unless in the constructor, which is not flexible enough. 

The current implementation caches `width` and `height` property of objects and modifications that affect size of object will lead to incorrect render result.

**Comment by qiumuyang** [comment]
Almost implemented by 5d6fff0.

TODO:

+ Modifying [Image](https://github.com/iori2333/render/blob/master/render/objects/image.py#L9) object by accessing ndarray `Image.im.base_im` will not be detected correctly.

  ```python
  from render import Alignment, Container, Image, Palette, RenderImage
  
  im = Image.from_color(100, 100, Palette.RED)
  container = Container.from_children([im])
  i1 = container.render()
  
  im.im.base_im[50:, 50:] = Palette.BLUE
  i2 = container.render()
  
  RenderImage.concat_horizontal([i1, i2], alignment=Alignment.CENTER).show()
  ```

  Consider to make `Image.im` a read-only property and can only be updated by assigning a new RenderImage to it.

+ Cacheable List and Dict do not support nesting.

**Comment by qiumuyang** [comment]
Closed in favor of 5d4467e & 6991cba
